### PR TITLE
13 18 getapi routers

### PR DESCRIPTION
**Pull request by Roothebear**
Included in this request:

- Add GET /api endpoint to return endpoints.json
- Refactor using express routers
- Tidy up testing suite adding more visible sections for error handling tests
- Add further tests for error handling DELETE comments by id.

Otherwise some minor changes to package.json and addition of files for hosting on heroku.

Thank you

### CODE REVIEW

File: models/articles.models.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
+const db = require("../db/connection.js");
+
+exports.selectArticles = (req) => {
+  // destructure optional query parameters, we will deal with each in turn
+  const { sort_by, order, topic } = req.query;
+
+  // set up array for query values used
+  const queryValues = [];
+  let queryStr = `SELECT articles.article_id, articles.title, articles.topic, articles.author, articles.body, articles.created_at, articles.votes, COUNT(comment_id )::INT AS comment_count
```
Comment (douglashellowell) [inline models/articles.models.js:9]: remember you can put strings using backticks on multiple lines!

might look nicer like this

```js
`SELECT 
   articles.article_id, 
   articles.title, 
   articles.topic, 
   articles.author, 
   articles.body, 
   articles.created_at, 
   articles.votes, 
   COUNT(comment_id )::INT AS comment_count 
  FROM articles 
  FULL JOIN comments 
    ON articles.article_id = comments.article_id`;
```

p.s you can also use `[table].*` to get all columns from one particular table! (i.e `articles.*`)

File: models/articles.models.js [hunk 2]
```diff
@@ -0,0 +1,151 @@
+const db = require("../db/connection.js");
+
+exports.selectArticles = (req) => {
+  // destructure optional query parameters, we will deal with each in turn
+  const { sort_by, order, topic } = req.query;
```
Comment (douglashellowell) [inline models/articles.models.js:5]: (subjective)

I would have the controller handle the `req` object and only pass down the values the model needs

File: models/articles.models.js [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  if (!(typeof topic === "undefined")) {
+    queryValues.push(topic);
+    queryStr += ` WHERE topic = $1`;
+  }
+
+  // deal with grouping
+  queryStr += ` GROUP BY articles.article_id`;
+
+  // deal with sort_by (alpha-numeric column to sort by)
+  if (!(typeof sort_by === "undefined")) {
```
Comment (douglashellowell) [inline models/articles.models.js:23]: can also use 

```js
if(sory_by)
```
or
```js
if(!!sort_by)
```
here

File: models/articles.models.js [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+  // add final ';'!
+  queryStr += `;`;
+
+  return db.query(queryStr, queryValues).then((result) => {
+    if (result.rows.length === 0) {
+      // as we have an empty array of articles in this case, check database for existence of the topic with further query
+      return db
+        .query("SELECT * FROM topics WHERE topics.slug = $1;", [topic])
+        .then((topics) => {
```
Comment (douglashellowell) [inline models/articles.models.js:60]: got a nested `.then` block here!

This whole check-if-topic-exists model could be extracted out

**Review by douglashellowell — CHANGES_REQUESTED** [review]
some refactoring of how you're passing `req` data down to your models needed!

File: models/articles.models.js (hunk 1 above)
Comment (Roothebear) [reply models/articles.models.js:9]: Thank you Doug, much nicer like that, also for the tip about table.* !

File: models/articles.models.js (hunk 2 above)
Comment (Roothebear) [reply models/articles.models.js:5]: I'll remember to ask about pros/cons when we catch up

File: models/articles.models.js (hunk 3 above)
Comment (Roothebear) [reply models/articles.models.js:23]: Okay that's great, I've seen that pop up a few times and need to review the booleans evaluated for this kind of thing.

File: models/articles.models.js (hunk 4 above)
Comment (Roothebear) [reply models/articles.models.js:60]: ah yes , this is a good time for me to return to the helper function and do a refactor. On the to do list.

**Comment by Roothebear** [comment]
Thank you for all the feedback Doug, really useful points to work on.

File: models/articles.models.js (hunk 2 above)
Comment (Roothebear) [reply models/articles.models.js:5]: Ok, I see what's happening here, got it, req could include lots we don't want!

Comment (Roothebear) [reply models/articles.models.js:5]: Refactor complete

**Pull request merged**